File: centralized-logging/scripts/logger_client.js
/**
 * logger_client.js — centralized logging client for Node.js apps.
 *
 * Talks to server.py over plain HTTP — no account, no API key. Zero
 * dependencies (uses the built-in `fetch`, Node 18+). Copy this file into
 * the target app. Batches events in memory and flushes on a timer, so a
 * log call never blocks the request/response path.
 *
 * Usage:
 *   const { CentralLogger } = require('./logger_client');
 *
 *   const logger = new CentralLogger({ appName: 'my-node-app' }); // defaults to http://127.0.0.1:4317
 *
 *   logger.info('server started', { context: { port: 3000 } });
 *
 *   try {
 *     await chargeCard(order);
 *   } catch (err) {
 *     logger.error(err, { context: { orderId: order.id }, requestId: req.id });
 *   }
 *
 *   process.on('SIGTERM', () => logger.flush());
 */

const os = require('os');
const crypto = require('crypto');

const DEFAULT_SERVER_URL = process.env.LOG_SERVER_URL || 'http://127.0.0.1:4317';

/**
 * Hash (errorType + message-with-digits-stripped) so repeats of the same
 * underlying bug collapse into one fingerprint instead of showing up as
 * separate 'distinct' errors just because an id in the message differs.
 */
function fingerprint(errorType, message) {
  const normalized = String(message || '').replace(/\d+/g, '#');
  const raw = `${errorType || ''}:${normalized}`;
  return crypto.createHash('sha256').update(raw).digest('hex').slice(0, 16);
}

class CentralLogger {
  /**
   * @param {object} opts
   * @param {string} opts.appName
   * @param {string} [opts.serverUrl] defaults to http://127.0.0.1:4317 (or LOG_SERVER_URL)
   * @param {string} [opts.environment='production']
   * @param {number} [opts.flushIntervalMs=2000]
   * @param {number} [opts.maxBatchSize=50]
   */
  constructor({
    appName,
    serverUrl = DEFAULT_SERVER_URL,
    environment = 'production',
    flushIntervalMs = 2000,
    maxBatchSize = 50,
  }) {
    this.appName = appName;
    this.environment = environment;
    this.endpoint = `${serverUrl.replace(/\/$/, '')}/logs`;
    this.host = os.hostname();
    this.maxBatchSize = maxBatchSize;
    this.queue = [];

    this._timer = setInterval(() => this._drainAndSend(), flushIntervalMs);
    this._timer.unref?.(); // don't keep the process alive just for this
  }

  // -- public logging API -----------------------------------------------

  debug(message, opts = {}) {
    this._enqueue('debug', message, opts);
  }

  info(message, opts = {}) {
    this._enqueue('info', message, opts);
  }

  warn(message, opts = {}) {
    this._enqueue('warn', message, opts);
  }

  /**
   * Accepts either a plain string message or an Error object — if given an
   * Error, the stack trace and error type are captured automatically.
   */
  error(messageOrError, opts = {}) {
    this._logErrorLike('error', messageOrError, opts);
  }

  fatal(messageOrError, opts = {}) {
    this._logErrorLike('fatal', messageOrError, opts);
  }

  // -- internals -----------------------------------------------------------

  _logErrorLike(level, messageOrError, opts) {
    if (messageOrError instanceof Error) {
      this._enqueue(level, messageOrError.message, {
        ...opts,
        errorType: messageOrError.name,
        stackTrace: messageOrError.stack,
      });
    } else {
      this._enqueue(level, messageOrError, opts);
    }
  }

  _enqueue(level, message, opts = {}) {
    const {
      errorType = null,
      stackTrace = null,
      context = {},
      requestId = null,
      sessionId = null,
      userRef = null,
      tags = [],
      durationMs = null,
      sourceFile = null,
      sourceLine = null,
      sourceFunction = null,
    } = opts;

    this.queue.push({
      app_name: this.appName,
      environment: this.environment,
      host: this.host,
      level,
      message,
      error_type: errorType,
      stack_trace: stackTrace,
      context,
      fingerprint: fingerprint(errorType, message),
      request_id: requestId,
      session_id: sessionId,
      user_ref: userRef,
      tags,
      duration_ms: durationMs,
      source_file: sourceFile,
      source_line: sourceLine,
      source_function: sourceFunction,
    });

    if (this.queue.length >= this.maxBatchSize) {
      this._drainAndSend();
    }
  }

  async _drainAndSend() {
    if (this.queue.length === 0) return;
    const batch = this.queue.splice(0, this.queue.length);
    try {
      await fetch(this.endpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(batch),
      });
    } catch {
      // Logging must never crash the app it's logging for — drop silently
      // if the log server isn't reachable rather than throw.
    }
  }

  /** Force an immediate send of whatever's queued. Call on shutdown. */
  async flush() {
    await this._drainAndSend();
  }
}

module.exports = { CentralLogger };

File: centralized-logging/scripts/logger_client.py
"""
logger_client.py — centralized logging client for Python apps.

Talks to server.py over plain HTTP — no account, no API key. Copy this file
into the target app. Batches events in memory and flushes them in the
background, so a log call never blocks the request/response path.

Usage:
    from logger_client import CentralLogger

    logger = CentralLogger(app_name="my-python-app")   # defaults to http://127.0.0.1:4317

    logger.info("worker started", context={"pid": os.getpid()})
    logger.warn("slow query", context={"duration_ms": 4200})

    try:
        risky_operation()
    except Exception:
        logger.exception("risky_operation failed", context={"job_id": job.id})

    logger.flush()   # optional — call on graceful shutdown to drain the queue
"""

from __future__ import annotations

import atexit
import hashlib
import json
import os
import queue
import re
import socket
import sys
import threading
import time
import traceback
from typing import Any, Optional

import requests

_DIGITS_RE = re.compile(r"\d+")
DEFAULT_SERVER_URL = os.environ.get("LOG_SERVER_URL", "http://127.0.0.1:4317")


def _fingerprint(error_type: Optional[str], message: str) -> str:
    """
    Hash (error_type + message-with-digits-stripped) so that, e.g.,
    "user 4821 not found" and "user 77 not found" collapse into the same
    fingerprint instead of showing up as two separate 'distinct' errors.
    """
    normalized = _DIGITS_RE.sub("#", message or "")
    raw = f"{error_type or ''}:{normalized}"
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()[:16]


class CentralLogger:
    """
    Batched, non-blocking logger that writes rows into the local log
    server (server.py) via a plain HTTP POST — no headers, no auth.
    """

    def __init__(
        self,
        app_name: str,
        server_url: str = DEFAULT_SERVER_URL,
        environment: str = "production",
        flush_interval_seconds: float = 2.0,
        max_batch_size: int = 50,
    ) -> None:
        self.app_name = app_name
        self.environment = environment
        self._endpoint = f"{server_url.rstrip('/')}/logs"
        self._host = socket.gethostname()
        self._queue: "queue.Queue[dict]" = queue.Queue()
        self._max_batch_size = max_batch_size
        self._flush_interval = flush_interval_seconds

        self._stop_event = threading.Event()
        self._worker = threading.Thread(target=self._run, daemon=True)
        self._worker.start()
        atexit.register(self.flush)

    # -- public logging API --------------------------------------------------

    def debug(self, message: str, **kwargs: Any) -> None:
        self._enqueue("debug", message, **kwargs)

    def info(self, message: str, **kwargs: Any) -> None:
        self._enqueue("info", message, **kwargs)

    def warn(self, message: str, **kwargs: Any) -> None:
        self._enqueue("warn", message, **kwargs)

    def error(self, message: str, **kwargs: Any) -> None:
        self._enqueue("error", message, **kwargs)

    def fatal(self, message: str, **kwargs: Any) -> None:
        self._enqueue("fatal", message, **kwargs)

    def exception(self, message: str, **kwargs: Any) -> None:
        """
        Call this from inside an `except:` block — captures the current
        exception's type and full traceback automatically.
        """
        exc_type, exc_value, exc_tb = sys.exc_info()
        error_type = exc_type.__name__ if exc_type else None
        stack_trace = "".join(traceback.format_exception(exc_type, exc_value, exc_tb))
        self._enqueue(
            "error",
            message,
            error_type=error_type,
            stack_trace=stack_trace,
            **kwargs,
        )

    # -- internals ------------------------------------------------------------

    def _enqueue(
        self,
        level: str,
        message: str,
        *,
        error_type: Optional[str] = None,
        stack_trace: Optional[str] = None,
        context: Optional[dict] = None,
        request_id: Optional[str] = None,
        session_id: Optional[str] = None,
        user_ref: Optional[str] = None,
        tags: Optional[list] = None,
        duration_ms: Optional[float] = None,
        source_file: Optional[str] = None,
        source_line: Optional[int] = None,
        source_function: Optional[str] = None,
    ) -> None:
        row = {
            "app_name": self.app_name,
            "environment": self.environment,
            "host": self._host,
            "level": level,
            "message": message,
            "error_type": error_type,
            "stack_trace": stack_trace,
            "context": context or {},
            "fingerprint": _fingerprint(error_type, message),
            "request_id": request_id,
            "session_id": session_id,
            "user_ref": user_ref,
            "tags": tags or [],
            "duration_ms": duration_ms,
            "source_file": source_file,
            "source_line": source_line,
            "source_function": source_function,
        }
        self._queue.put(row)

    def _run(self) -> None:
        """Background thread: batch up queued rows and flush periodically."""
        while not self._stop_event.is_set():
            time.sleep(self._flush_interval)
            self._drain_and_send()

    def _drain_and_send(self) -> None:
        batch = []
        while len(batch) < self._max_batch_size:
            try:
                batch.append(self._queue.get_nowait())
            except queue.Empty:
                break
        if not batch:
            return
        try:
            requests.post(self._endpoint, json=batch, timeout=5)
        except requests.RequestException:
            # Logging must never crash the app it's logging for. If the log
            # server isn't running, drop the batch silently rather than raise.
            pass

    def flush(self) -> None:
        """Force an immediate send of whatever's queued. Call on shutdown."""
        self._drain_and_send()
